asset-transfer-private-data/chaincode-go/privateAssetTransfer.go

```diff
@@ -0,0 +1,469 @@
+	}
+	assetJSONasBytes, err := json.Marshal(asset)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	// Save asset to private data collection
+	err = ctx.GetStub().PutPrivateData(assetCollection, assetInput.ID, assetJSONasBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put Asset: %s", err.Error())
+	}
+
+	// Save asset details to collection visible to owning organization
+	assetPrivateDetails := &AssetPrivateDetails{
+		ID:       			 	assetInput.ID,
+		AppraisedValue:   assetInput.AppraisedValue,
+	}
+
+	assetPrivateDetailsAsBytes, err := json.Marshal(assetPrivateDetails) // marshal asset details to JSON
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+
+	// get collection name for this organization
+	orgCollection, err := getCollectionName(ctx)
+
+	// put asset into owners org specific private data collection
+	err = ctx.GetStub().PutPrivateData(orgCollection, assetInput.ID, assetPrivateDetailsAsBytes)
+	if err != nil {
+		return fmt.Errorf("failed to put asset private details: %s", err.Error())
+	}
+	return nil
+}
+
+// AgreeToPrice is used by the potential buyer of the asset to agree to the
+// asset price. The agreed to appraisal value is stored in the buying orgs
+// org specifc collection, while the the buyer client ID is stored in the asset collection
+// using a composite key
+func (s *SmartContract) AgreeToPrice(ctx contractapi.TransactionContextInterface, assetID string) error {
+
+	// Get ID of submitting client identity
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return fmt.Errorf("failed to get verified OrgID: %s", err.Error())
+	}
+
+	// price is private, therefore it gets passed in transient field
+	transMap, err := ctx.GetStub().GetTransient()
+	if err != nil {
+		return fmt.Errorf("Error getting transient: " + err.Error())
+	}
+
+	// persist the JSON bytes as-is so that there is no risk of nondeterministic marshaling.
+	priceJSON, ok := transMap["asset_price"]
+	if !ok {
+		return fmt.Errorf("asset_price key not found in the transient map")
+	}
+
+	// get collection name for this organization
+	orgCollection, err := getCollectionName(ctx)
+
+	// put price in the org specifc private data collection
+	err = ctx.GetStub().PutPrivateData(orgCollection, assetID, priceJSON)
+	if err != nil {
+		return fmt.Errorf("failed to put asset bid: %s", err.Error())
+	}
+
+	// create agreeement where you indicate the identity has agreed to purchase
+	transferAgreeKey, err := ctx.GetStub().CreateCompositeKey("transferAgreement",[]string{assetID})
+	if err != nil {
+		return fmt.Errorf("failed to create composite key: %s", err.Error())
+	}
+
+	err = ctx.GetStub().PutPrivateData(assetCollection, transferAgreeKey, []byte(clientID))
+	if err != nil {
+		return fmt.Errorf("failed to put asset bid: %s", err.Error())
+	}
+
+	return nil
+}
+
+
+// transfer a asset by setting a new owner ID on the asset
+func (s *SmartContract) TransferAsset(ctx contractapi.TransactionContextInterface) error {
```

Comment:
as in the ledger_query go chaincode, the function should be
`func (t *SimpleChaincode) TransferAsset(ctx contractapi.TransactionContextInterface, assetID, newOwner string)`

Reply:
as a developer who starts from basic, to ledger-queries, I was expecting TransferAsset to have same signature and semantics (transfer a specific assetID to newOwner). If there is nothing inherently wrong in calling TransferAsset with a assetID param, in priv data, we should demonstrate the same.  I think.

Reply:
we should continue demonstrating transient field usage. But as a new reader , this code pattern, is a bit misleading I think. A developer may associate that private-collection and transient args (ctx.GetStub().GetTransient()) , are to be used together, strictly together. 

If we are not constrained to stick to previous marble_02 pattern, we can
- use `AssetPrivateDetails` as transient args to a func like AgreeToTransfer
- keep the transfer itself public: the assetID, & new_owner args public on the ledger/channel (& data in the collection for org1 & org2) << close to many real life scenarios
- that will keep  this func signature same as basic & ledger_query sample `TransferAsset(ctx contractapi.TransactionContextInterface, assetID, newOwner string)`  <<  minimize cognitive overload

Reply:
Asset are visible to all organizations in the collection, but they are not 'public' on the channel. Specifically we want hide the data from ordering service in this scenario.
We also want to demonstrate multi-org collections and org-specific collections.
We also want to demonstrate that some collections have an endorsement policy that overrides the chaincode default, and some don't.
So I think we need to keep the collections as-is.

Reply:
I updated the chaincode so that all values are passed through transient data.